planner: add variables for instance plan cache (#18830)
pingcap/docs: system-variables.md

- Default value: `OFF`
- This variable controls whether Index Join is supported when the inner table has `Selection` or `Projection` operators on it. The default value `OFF` means that Index Join is not supported in this scenario.

### tidb_enable_instance_plan_cache <span class="version-mark">New in v8.4.0</span>

> **Warning:**
>
> Currently, Instance Plan Cache is an experimental feature. It is not recommended that you use it in the production environment. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.

- Scope: GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
- Type: Boolean
- Default value: `OFF`

> - This variable controls whether to enable the Instance Plan Cache feature.

Reviewer: ```suggestion
- This variable controls whether to enable the Instance Plan Cache feature. This feature implements instance-level execution plan cache, which allows  all sessions within the same TiDB instance to share the execution plan cache, thereby improving memory utilization.
```

> - This variable controls whether to enable the Instance Plan Cache feature. This feature implements instance-level execution plan cache, which allows  all sessions within the same TiDB instance to share the execution plan cache, thereby improving memory utilization.

Reviewer 2: ```suggestion
- This variable controls whether to enable the Instance Plan Cache feature. This feature implements instance-level execution plan cache, which allows  all sessions within the same TiDB instance to share the execution plan cache, thereby improving memory utilization. It's highly recommended to disable instance at session level before activating this feature.  Please refer to [SQL Prepared Execution Plan Cache](/sql-prepared-plan-cache.md) and [SQL Non-Prepared Execution Plan Cache](/sql-non-prepared-plan-cache.md) for details. 
```

- Unit: Rows
- This variable is used to set the number of rows for the initial chunk during the execution process. The number of rows for a chunk directly affects the amount of memory required for a single query. You can roughly estimate the memory needed for a single chunk by considering the total width of all columns in the query and the number of rows for the chunk. Combining this with the concurrency of the executor, you can make a rough estimation of the total memory required for a single query. It is recommended that the total memory for a single chunk does not exceed 16 MiB.

> ### tidb_instance_plan_cache_target_mem_size <span class="version-mark">New in v8.4.0</span>

Author: ### tidb_instance_plan_cache_reserved_percentage <span class="version-mark">New in v8.4.0</span>

> **Warning:**
>
> Currently, Instance Plan Cache is an experimental feature. It is not recommended that you use it in the production environment. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.

- Scope: GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
- Type: Float
- Default value: `0.1`
- Range: `[0, 1]`
- This variable controls the amount of memory reserved for each eviction. TiDB starts eviction when the Plan Cache is full, which means its memory usage reaches `tidb_instance_plan_cache_max_size`. Each time TiDB evicts `tidb_instance_plan_cache_max_size * tidb_instance_plan_cache_reserved_percentage` bytes.

> ### tidb_instance_plan_cache_target_mem_size <span class="version-mark">New in v8.4.0</span>
> 
> > **Warning:**
> >
> > Currently, Instance Plan Cache is an experimental feature. It is not recommended that you use it in the production environment. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.
> 
> - Scope: GLOBAL
> - Persists to cluster: Yes
> - Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
> - Type: Integer
> - Default value: `104857600` (which is 100 MiB)
> - This variable sets the target memory size for Instance Plan Cache. TiDB periodically cleans up the Instance Plan Cache in the background to ensure its memory usage does not exceed the value of this variable.

Author: ```suggestion
### tidb_instance_plan_cache_reserved_percentage <span class="version-mark">New in v8.4.0</span>

> **Warning:**
>
> Currently, Instance Plan Cache is an experimental feature. It is not recommended that you use it in the production environment. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.

- Scope: GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
- Type: Float
- Default value: `0.1`
- Range: `[0, 1]`
- This variable controls the amount of memory reserved for each eviction. TiDB starts eviction when the Plan Cache is full, which means its memory usage reaches `tidb_instance_plan_cache_max_size`. Each time TiDB evicts `tidb_instance_plan_cache_max_size * tidb_instance_plan_cache_reserved_percentage` bytes.
```

- Unit: Rows
- This variable is used to set the number of rows for the initial chunk during the execution process. The number of rows for a chunk directly affects the amount of memory required for a single query. You can roughly estimate the memory needed for a single chunk by considering the total width of all columns in the query and the number of rows for the chunk. Combining this with the concurrency of the executor, you can make a rough estimation of the total memory required for a single query. It is recommended that the total memory for a single chunk does not exceed 16 MiB.

### tidb_instance_plan_cache_reserved_percentage <span class="version-mark">New in v8.4.0</span>

> **Warning:**
>
> Currently, Instance Plan Cache is an experimental feature. It is not recommended that you use it in the production environment. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.

- Scope: GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
- Type: Float
- Default value: `0.1`
- Range: `[0, 1]`

> - This variable controls the amount of memory reserved for each eviction. TiDB starts eviction when the Plan Cache is full, which means its memory usage reaches `tidb_instance_plan_cache_max_size`. Each time TiDB evicts `tidb_instance_plan_cache_max_size * tidb_instance_plan_cache_reserved_percentage` bytes.

Reviewer 2: ```suggestion
- This variable controls the amount of free memory reserved after eviction takes place. TiDB starts evicting execution plans when the total memory used by plan cache reaches `tidb_instance_plan_cache_max_size`. The eviction stops in case the percentage of free memory is larger than the value defined by `tidb_instance_plan_cache_reserved_percentage`.

### tidb_instance_plan_cache_max_size <span class="version-mark">New in v8.4.0</span>
> **Warning:**
>
> Currently, Instance Plan Cache is an experimental feature. is an experimental feature. It is not recommended that you use it in the production environment. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.
- Scope: GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
- Type: Integer
- Default value: `125829120` (which is 120 MiB)

> - This variable sets the maximum memory usage for Instance Plan Cache.

Reviewer 2: ```suggestion
- This variable sets the maximum memory usage in bytes for Instance Plan Cache.
```

- Unit: Rows
- This variable is used to set the number of rows for the initial chunk during the execution process. The number of rows for a chunk directly affects the amount of memory required for a single query. You can roughly estimate the memory needed for a single chunk by considering the total width of all columns in the query and the number of rows for the chunk. Combining this with the concurrency of the executor, you can make a rough estimation of the total memory required for a single query. It is recommended that the total memory for a single chunk does not exceed 16 MiB.

### tidb_instance_plan_cache_target_mem_size <span class="version-mark">New in v8.4.0</span>

> **Warning:**
>
> Currently, Instance Plan Cache is an experimental feature. It is not recommended that you use it in the production environment. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.

- Scope: GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
- Type: Integer
- Default value: `104857600` (which is 100 MiB)

> - TiDB clean the Instance Plan Cache periodically in the background, and this variable controls the target memory size of the cache, which means TiDB tries to clean the plan cache to this memory size each time.

Reviewer: ```suggestion
- This variable sets the target memory size for Instance Plan Cache. TiDB periodically cleans up the Instance Plan Cache in the background to ensure its memory usage does not exceed the value of this variable.
```

- Unit: Rows
- This variable is used to set the number of rows for the initial chunk during the execution process. The number of rows for a chunk directly affects the amount of memory required for a single query. You can roughly estimate the memory needed for a single chunk by considering the total width of all columns in the query and the number of rows for the chunk. Combining this with the concurrency of the executor, you can make a rough estimation of the total memory required for a single query. It is recommended that the total memory for a single chunk does not exceed 16 MiB.

### tidb_instance_plan_cache_reserved_percentage <span class="version-mark">New in v8.4.0</span>

> **Warning:**
>
> Currently, Instance Plan Cache is an experimental feature. It is not recommended that you use it in the production environment. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.

- Scope: GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
- Type: Float
- Default value: `0.1`
- Range: `[0, 1]`

> - This variable controls the percentage of idle memory for Instance Plan Cache after memory eviction. When the memory used by Instance Plan Cache reaches the limit set by [`tidb_instance_plan_cache_max_size`](#tidb_instance_plan_cache_max_size-new-in-v840), TiDB starts evicting execution plans from memory using the Least Recently Used (LRU) algorithm until the idle memory percentage exceeds the value set by [`tidb_instance_plan_cache_reserved_percentage`](#tidb_instance_plan_cache_reserved_percentage-new-in-v840).

Reviewer: ```suggestion
- This variable controls the percentage of idle memory for [Instance Plan Cache](#tidb_enable_instance_plan_cache-new-in-v840) after memory eviction. When the memory used by Instance Plan Cache reaches the limit set by [`tidb_instance_plan_cache_max_size`](#tidb_instance_plan_cache_max_size-new-in-v840), TiDB starts evicting execution plans from memory using the Least Recently Used (LRU) algorithm until the idle memory percentage exceeds the value set by [`tidb_instance_plan_cache_reserved_percentage`](#tidb_instance_plan_cache_reserved_percentage-new-in-v840).
```

> - This variable controls the percentage of idle memory for [Instance Plan Cache](#tidb_enable_instance_plan_cache-new-in-v840) after memory eviction. When the memory used by Instance Plan Cache reaches the limit set by [`tidb_instance_plan_cache_max_size`](#tidb_instance_plan_cache_max_size-new-in-v840), TiDB starts evicting execution plans from memory using the Least Recently Used (LRU) algorithm until the idle memory percentage exceeds the value set by [`tidb_instance_plan_cache_reserved_percentage`](#tidb_instance_plan_cache_reserved_percentage-new-in-v840).

Reviewer 3: ```suggestion
- This variable controls the percentage of idle memory reserved for [Instance Plan Cache](#tidb_enable_instance_plan_cache-new-in-v840) after memory eviction. When the memory used by Instance Plan Cache reaches the limit set by [`tidb_instance_plan_cache_max_size`](#tidb_instance_plan_cache_max_size-new-in-v840), TiDB starts evicting execution plans from memory using the Least Recently Used (LRU) algorithm until the idle memory percentage exceeds the value set by [`tidb_instance_plan_cache_reserved_percentage`](#tidb_instance_plan_cache_reserved_percentage-new-in-v840).
```

### tidb_instance_plan_cache_max_size <span class="version-mark">New in v8.4.0</span>

> **Warning:**
>
> Currently, Instance Plan Cache is an experimental feature. is an experimental feature. It is not recommended that you use it in the production environment. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.

- Scope: GLOBAL
- Persists to cluster: Yes
- Applies to hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value): No
- Type: Integer
- Default value: `125829120` (which is 120 MiB)
- Unit: Bytes

> - This variable sets the maximum memory usage for Instance Plan Cache.

Reviewer: ```suggestion
- This variable sets the maximum memory usage for [Instance Plan Cache](#tidb_enable_instance_plan_cache-new-in-v840).
```
